### PR TITLE
Support for hot-pluggable wifi interfaces

### PR DESCRIPTION
**Pull request by layereight**
* 2 optional configs for interfaces_ether_interfaces
  * allowclass: anything other than 'auto'
  * wpa-conf: wpa_supplicant configuration
* add "--allow auto" flag to ifup/ifdown to only bounce 'auto' interfaces

### CODE REVIEW

File: handlers/main.yml [hunk 1]
```diff
@@ -95,11 +95,11 @@
     nohup bash -c "
     returncode=0
     {% for interface in all_interfaces_changed | reverse %}
-    ifdown {{ interface }};
+    ifdown --allow auto {{ interface }};
```
Comment (markgoddard) [inline handlers/main.yml]: This won't work on RHEL-derived systems - there ifdown/ifup only takes one argument, the interface name. Could you add a short comment to explain what this is for? I understand myself now but had to read a few man pages first :)

Comment (layereight) [reply handlers/main.yml]: Hmm, I see.

On debian system the `--allow` flag to ifup/ifdown behaves like a filter to the interfaces configured in `/etc/network/interfaces`. Typically you'd configure an interface there like `auto eth0`, meaning the state of the interface is managed by ifup/ifdown. 

Not so for hot-pluggable wifi interfaces (e.g. usb wifi dongle). Here you would configure them like `allow-hotplug wlan0`. This means the interface's state is managed by the kernel's hotplug events. So if you plug the wifi dongle in, the interface will be brought up and configured (static IP or dhcp or whatever). If you unplug the wifi dongle, the interface will be brought down. Trying to manage a hot-pluggable interface using ifup/ifdown will just fail.

Long story short: On Debian you cannot bounce an interface marked with `allow-hotplug` using ifup/ifdown. That's why I added the `--allow auto` argument to ifup/ifdown. This will only be effective for interface marked with `auto`. Interface marked with `allow-hotplug` will simply be ignored then.

On a debian system also see:
```
$ man interfaces
$ man ifup
```

@markgoddard I can implement a conditional switch here for RHEL systems. Or do you see another way?

Comment (markgoddard) [reply handlers/main.yml]: Thanks, seems like a reasonable solution to the problem. How about this (untested):

`ifup {% if ansible_os_family == 'Debian' %}--allow auto {% endif %}{{ interface }};`

Comment (markgoddard) [reply handlers/main.yml]: Also, could you put a brief explanation as a comment in the code?

Comment (layereight) [reply handlers/main.yml]: @markgoddard Looks good. Give me some time for the change and to test. Might only be at the weekend.

Btw, do you see any possibility to do a negative test for RHEL on Travis. Travis is only using Ubuntu after all. :-/

Comment (markgoddard) [reply handlers/main.yml]: It would be possible by using a RHEL docker container. It would be a bit fiddly though.

**Comment by layereight** [comment]
@markgoddard BTW, could you trigger a reimport of the project in ansible galaxy? You can also automate this with a Travis notification. See http://docs.ansible.com/ansible/latest/galaxy.html#travis-integrations

Maybe introducing tags for versioning releases is also a good idea. So people can explicitly stick to a certain version and upgrade the role in their playbooks when they feel like it.

**Comment by markgoddard** [comment]
@layereight I've not needed to do a reimport - galaxy seems to pick up the new code automatically. I'll look into it in case I've just been lucky in my timing.

@michaelrigart what do you think about versioning?

**Comment by layereight** [comment]
@markgoddard https://galaxy.ansible.com/MichaelRigart/interfaces/ shows me last successful import was 10 month ago (or am I looking at the wrong role). Please compare with the imports from my fork: https://galaxy.ansible.com/layereight/interfaces/

You can also see that galaxy picks up the tags from the repos as version. In my playbook I have a **roles.yml** pointing the roles to a specific version:

```yml
  - name: layereight.interfaces
    src: layereight.interfaces
    version: "1.6"
```
To install the roles I do:
```
$ ansible-galaxy install -r roles.yml
```
See http://docs.ansible.com/ansible/latest/galaxy.html#installing-multiple-roles-from-a-file

**Comment by markgoddard** [comment]
@layereight the main project in which I use this role is https://github.com/stackhpc/kayobe, where I have a requirements.yml file like your roles.yml. I tend to omit the version, and pull down the latest:

`- src: MichaelRigart.interfaces`

This seems to always pull down the master branch from github, despite the lack of imports on galaxy, perhaps because there are no tags.

Let's see what @michaelrigart says re versioning - I can see the benefit of being able to specify a version but it does add a little maintenance overhead.

**Comment by layereight** [comment]
@markgoddard Thanks! I appreciate your consideration.

**Comment by michaelrigart** [comment]
@markgoddard @layereight 

Versioning sounds like a good plan. I just re-imported the repo in galaxy.
Any thoughts on versioning schema, without creating too much overhead ( <major>.<minor>.<bugfix> ) ?

**Comment by markgoddard** [comment]
Yeah, https://semver.org/ seems like the way to go.

**Comment by markgoddard** [comment]
@michaelrigart I think you would need to setup the travis notification as it's your account.

**Comment by layereight** [comment]
@michaelrigart Nice, thanks!

The good thing about versioning schemas is that there are so many you can choose from. ;-) I usually keep it simple. 'major'.'minor'.'bugfix' is totally fine. Important thing is that you have versioning. The big advantage for people is that they can stick to a version and then adapt to a new version (with a change potentially breaking their playbook) at their own pace.

**Comment by michaelrigart** [comment]
@markgoddard : what kind of notifications do you mean? Aren't notifications configured through the travis-ci.yml file?

**Comment by markgoddard** [comment]
@michaelrigart yes they are - http://docs.ansible.com/ansible/latest/galaxy.html#travis-integrations. I think it requires the role owner's account to generate an appropriate token for the notification to use though.

**Comment by michaelrigart** [comment]
@markgoddard : ok, integration should be ready. Can you try adding the notification entry in the yml?

```Added integration for travis michaelrigart/ansible-role-interfaces```

**Comment by layereight** [comment]
@michaelrigart Btw, are you going to merge my PR? ;-)

**Comment by michaelrigart** [comment]
@markgoddard @layereight I'm going to open an issue tomorrow regarding the versioning request.

**Comment by layereight** [comment]
@michaelrigart Cool, thanks!
